Sources/AsyncHTTPClient/Utils.swift

```diff
@@ -47,21 +53,85 @@ public final class HTTPClientCopyingDelegate: HTTPClientResponseDelegate {
 }
 
 extension ClientBootstrap {
-    static func makeHTTPClientBootstrapBase(group: EventLoopGroup, host: String, port: Int, configuration: HTTPClient.Configuration, channelInitializer: ((Channel) -> EventLoopFuture<Void>)? = nil) -> ClientBootstrap {
-        return ClientBootstrap(group: group)
-            .channelOption(ChannelOptions.socket(SocketOptionLevel(IPPROTO_TCP), TCP_NODELAY), value: 1)
+    fileprivate func makeClientTCPBootstrap(
+        host: String,
+        requiresTLS: Bool,
+        configuration: HTTPClient.Configuration
+    ) throws -> NIOClientTCPBootstrap {
+        // if there is a proxy don't create TLS provider as it will be added at a later point
+        if configuration.proxy != nil {
+            return NIOClientTCPBootstrap(self, tls: NIOInsecureNoTLS())
+        } else {
+            let tlsConfiguration = configuration.tlsConfiguration ?? TLSConfiguration.forClient()
+            let sslContext = try NIOSSLContext(configuration: tlsConfiguration)
+            let hostname = (!requiresTLS || host.isIPAddress) ? nil : host
+            let tlsProvider = try NIOSSLClientTLSProvider<ClientBootstrap>(context: sslContext, serverHostname: hostname)
+            return NIOClientTCPBootstrap(self, tls: tlsProvider)
+        }
+    }
+}
+
+extension NIOClientTCPBootstrap {
+    /// create a TCP Bootstrap based off what type of `EventLoop` has been passed to the function.
+    fileprivate static func makeBootstrap(
+        on eventLoop: EventLoop,
+        host: String,
+        requiresTLS: Bool,
+        configuration: HTTPClient.Configuration
+    ) throws -> NIOClientTCPBootstrap {
+        let bootstrap: NIOClientTCPBootstrap
+        #if canImport(Network)
+            // if eventLoop is compatible with NIOTransportServices create a NIOTSConnectionBootstrap
+            if #available(OSX 10.14, iOS 12.0, tvOS 12.0, watchOS 6.0, *),
```

Comment:
this doesn't look quite right. If somebody passes you a `MultiThreadedEventLoopGroup` on macOS, even if Network.framework is supported, you'll have to use `ClientBootstrap`.

In other words, here we should just try all the options, one after the other using `validatingGroup:`. Alternatively, you can store what kind of ELG it is and switch over it here.

The problem is that the user can do (on new macOSes):

```
let elg = MultiThreadedELG(...)
HTTPClient(eventLoopGroup: .shared(elg))
```

and that needs to remain valid.

Reply:
err, we are trying each option one after the other. If attempts to create an `NIOTSConnectionBootstrap` and if that fails it attempts to create a `ClientBootstrap`

Reply:
sorry @adam-fowler ! I didn't see that the `let tsBootstrap = ...` is part of the `if` above! Ignore this

Reply:
maybe an extra empty line after `let tsBootstrap = ...` ?

```
                let tsBootstrap = NIOTSConnectionBootstrap(validatingGroup: eventLoop) {
                let tlsConfiguration = configuration.tlsConfiguration ?? TLSConfiguration.forClient()
```

looks like two normal `let`s...

Reply:
didn't see the `{` and the `,` :)

Reply:
I'll stick the { on a newline

Reply:
@adam-fowler you might get into trouble with SwiftFormat :P. If it works, then cool, if not, let's just disregard this comment.

Reply:
I'll run swift format on it and see. I think one of the versions of swift format was actually encouraging this. Maybe the Apple one, which you don't use ..?!?

Reply:
Yep swift format didn't like it. I've re-formatted it slightly. Also moved the tlsConfiguration down slightly as it doesn't need initialised if we have a proxy.